### PR TITLE
remove support for dune < 2.4

### PR DESCRIPTION
**Pull request by andlaus**
this allows quite a substantial amount of code decluttering, mainly because Dune 2.4 renamed mapper.map() to mapper.index().

### CODE REVIEW

**Comment by andlaus** [comment]
jenkins build this with downstreams please

**Comment by alfbr** [comment]
This is a rather invasive change. We are still not able to set up a stable build chain without 2.3.1 on Debian stable. Our own Jenkins server runs the old stable and just shifted from 2.3.1 this week. We are working hard on making that happen for linuxbenchmarking.com too, but unsuccessful so far. Hence, this PR is a bit too early, and we need at least some more weeks before we kill off 2.3.1 support. Preferably we do it after vacation time.

**Comment by andlaus** [comment]
>  Our own Jenkins server runs the old stable and just shifted from 2.3.1 this week.

@akva2 recently announced on the OPM mailing that he successfully upgraded Jenkins to Dune 2.4.1. Unfortunately I cannot provide a link because the archiving infrastructure of the mailing list seems to be broken at the moment.

**Comment by akva2** [comment]
... guess on who's instructions ;)

**Comment by andlaus** [comment]
and, btw, for me as the eWoms maintainer, it is a considerable pain in the b*** to continue maintaining Dune 2.3 compatibility in the code, and it is an even more considerable pain to keep the unit tests passing. that does not even touch on the fact that the performance on dune 2.3 is suboptimal -- at least I've been told so.

tl;dr: we should not encourage people to shoot themselves into their feet.

**Comment by andlaus** [comment]
Okay, I've also adapted the remaining modules; at least the ones which were part of the 2017.04 release:

- OPM/opm-material#238
- OPM/opm-grid#277
- OPM/opm-core#1170
- OPM/opm-simulators#1241
- OPM/opm-upscaling#220

opm-common, opm-parser and opm-output do not use dune and thus don't need to be adapted. travis still needs to be dealt with one way or another. (probably we should think about removing it for everything except opm-parser.)

**Comment by alfbr** [comment]
> for me as the eWoms maintainer, it is a considerable pain in the b*** 

Yes, and that is why we are working very hard to make the transition fast. All I am saying is that we are not there just yet, and as long as eWoms is in the build chain of flow, this is probably something you need to deal with, and accept the need for coordination.

**Comment by andlaus** [comment]
thanks for getting this in!

**Comment by alfbr** [comment]
You are welcome :-)
